model/log/sender.go

```diff
@@ -0,0 +1,210 @@
+	// location of the logger output.
+	Local send.Sender
+	// MaxBufferSize is the maximum number of bytes to buffer before
+	// persisting log data. Defaults to 10MB.
+	MaxBufferSize int
+	// FlushInterval is time interval at which to flush log lines,
+	// regardless of whether the max buffer size has been reached. A flush
+	// interval less than or equal to 0 will disable timed flushes.
+	FlushInterval time.Duration
+}
+
+// sender implements the send.Sender interface for persisting Evergreen logs.
+type sender struct {
+	mu         sync.Mutex
+	ctx        context.Context
+	cancel     context.CancelFunc
+	opts       LoggerOptions
+	write      logWriter
+	buffer     []LogLine
+	bufferSize int
+	lastFlush  time.Time
+	timer      *time.Timer
+	closed     bool
+	*send.Base
+}
+
+type logWriter func(context.Context, []LogLine) error
+
+// makeLogger returns a sender backed by the Evergreen log service.
+func makeLogger(ctx context.Context, name string, opts LoggerOptions, write logWriter) (send.Sender, error) { //nolint: unused
+	ctx, cancel := context.WithCancel(ctx)
+	s := &sender{
+		ctx:    ctx,
+		cancel: cancel,
+		opts:   opts,
+		write:  write,
+		Base:   send.NewBase(name),
+	}
+
+	if err := s.SetErrorHandler(send.ErrorHandlerFromSender(s.opts.Local)); err != nil {
+		return nil, errors.Wrap(err, "setting default error handler")
+	}
+
+	if opts.MaxBufferSize <= 0 {
+		opts.MaxBufferSize = defaultMaxBufferSize
+	}
+
+	if opts.FlushInterval > 0 {
+		go s.timedFlush()
+	}
+
+	return s, nil
+}
+
+// Send sends the given message to the Evergreen log service. This function
+// buffers the messages until the maximum allowed buffer size is reached, at
+// which point the messages in the buffer are written to persistent storage by
+// the backing log service. Send is thread safe.
+func (s *sender) Send(m message.Composer) {
+	ts := time.Now().UnixNano()
+
+	if !s.Level().ShouldLog(m) {
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
+		s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.New("cannot call Send on a closed sender")))
+		return
+	}
+
+	for _, line := range strings.Split(m.String(), "\n") {
+		if line == "" {
+			continue
+		}
+
+		logLine, err := s.opts.Parse(line)
+		if err != nil {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Wrap(err, "parsing log line")))
+			return
+		}
+		if logLine.Priority == 0 {
+			logLine.Priority = m.Priority()
+		}
+		if !logLine.Priority.IsValid() {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Errorf("invalid log line priority %d", logLine.Priority)))
+			return
+		}
+		if logLine.Timestamp == 0 {
+			logLine.Timestamp = ts
+		}
+		if logLine.Timestamp < 0 {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Errorf("invalid log line timestamp %d", logLine.Timestamp)))
+			return
+		}
+
+		s.buffer = append(s.buffer, logLine)
+		s.bufferSize += len(line)
+		if s.bufferSize > s.opts.MaxBufferSize {
```

Comment:
maybe we'd want `>=`?

Reply:
It doesn't actually matter, either way is technically correct. This just makes it a bit easier to test "overflowing" the buffer.